907 Added elisp function call for run, test, compile commands.

maximvl opened the pull request:
I've added check for string type for command read, cause in other case we should somehow detect if user's input is string compilation command or elisp function, so user input will only be requested if project's compilation command is a string.

bbatsov commented on projectile.el:
```diff
@@ -2201,7 +2202,9 @@ with a prefix ARG."
-    (compilation-start compilation-cmd)))
+    (if (symbolp compilation-cmd)
+        (funcall compilation-cmd)
+        (compilation-start compilation-cmd))))
```
This is not indented properly.

bbatsov commented on projectile.el:
```diff
@@ -2233,13 +2236,16 @@ with a prefix ARG."
-    (compilation-start test-cmd)))
+    (if (symbolp test-cmd)
+        (funcall test-cmd)
+        (compilation-start test-cmd))))
```
This as well.

bbatsov commented on projectile.el:
```diff
@@ -2250,13 +2256,16 @@ with a prefix ARG."
-    (compilation-start run-cmd)))
+    (if (symbolp run-cmd)
+        (funcall run-cmd)
+        (compilation-start run-cmd))))
```
And this.

bbatsov commented:
Apart from my inline comments:

* the commit message should match the project's style - `[Fix #...] Add ability to ...`.
* this should be mentioned in the readme & changelog.

maximvl commented:
@bbatsov Allright. I've fixed indentation, commit message, added readme and changelog mention.
I've also fixed code to check with `functionp`, cause `lambda` and `nil` expression could cause problems with `symbolp`.

bbatsov commented on projectile.el:
```diff
@@ -2192,7 +2192,8 @@ with a prefix ARG."
          (default-directory (or dir (projectile-compilation-dir)))
          (default-cmd (projectile-compilation-command default-directory))
-         (compilation-cmd (if (or compilation-read-command arg)
+         (compilation-cmd (if (and (stringp default-cmd)
```
Btw, what about the `nil` case? Should people be prompted there or not?

bbatsov commented on README.md:
```diff
@@ -536,6 +536,7 @@ There are a few variables that are intended to be customized via `.dir-locals.el
 
 They're all set to `nil` by default, but by setting them you'll override the
 default commands per each supported project type.
+These variables can be strings to run external commands or elisp functions.  
```
elisp -> Emacs Lisp

Perhaps we should also have 1-2 examples here for clarity's sake.

bbatsov commented on projectile.el:
```diff
@@ -2233,13 +2236,16 @@ with a prefix ARG."
   (let* ((project-root (projectile-project-root))
          (default-cmd (projectile-test-command project-root))
-         (test-cmd (if (or compilation-read-command arg)
+         (test-cmd (if (and (stringp default-cmd)
```
This check repeats a few times, maybe it should be extracted/abstracted away.

maximvl commented:
Ok, I've added `projectile-maybe-read-command` to abstract command reading, and `projectile-run-compilation` for compilation command checks.

bbatsov commented on README.md:
```diff
@@ -536,6 +536,10 @@ There are a few variables that are intended to be customized via `.dir-locals.el
 
 They're all set to `nil` by default, but by setting them you'll override the
 default commands per each supported project type.
+These variables can be strings to run external commands or Emacs lisp functions:
```
lisp -> Lisp

bbatsov commented on README.md:
````diff
@@ -536,6 +536,10 @@ There are a few variables that are intended to be customized via `.dir-locals.el
 They're all set to `nil` by default, but by setting them you'll override the
 default commands per each supported project type.
+These variables can be strings to run external commands or Emacs lisp functions:
+```
````
There should be a blank line before the code block.

bbatsov commented on README.md:
````diff
@@ -536,6 +536,10 @@ There are a few variables that are intended to be customized via `.dir-locals.el
 default commands per each supported project type.
+These variables can be strings to run external commands or Emacs lisp functions:
+```
+(setq projectile-test-cmd 'custom-test-function)
````
' -> #'

bbatsov commented on projectile.el:
```diff
@@ -2182,6 +2182,17 @@ Should be set via .dir-locals.el.")
                (file-name-as-directory projectile-project-compilation-dir)))
     (projectile-project-root)))
 
+(defun projectile-maybe-read-command (arg default-cmd prompt)
```
Add a docstring here.

bbatsov commented on projectile.el:
```diff
@@ -2182,6 +2182,17 @@ Should be set via .dir-locals.el.")
+      (projectile-read-command prompt default-cmd)
+    default-cmd))
+
+(defun projectile-run-compilation (cmd)
```
Add a docstring here.

bbatsov commented:
The code looks good, apart from a few tiny remarks I've added. Address them and remove the `.` in the commit message's title (titles are not complete sentences after all).

maximvl commented:
Done, should be fine now.
